metaflow/plugins/deployer.py

```diff
@@ -0,0 +1,206 @@
+
+
+def handle_timeout(tfp_runner_attribute, command_obj: CommandManager):
+    try:
+        content = read_from_file_when_ready(tfp_runner_attribute.name, timeout=10)
+        return content
+    except TimeoutError as e:
+        stdout_log = open(command_obj.log_files["stdout"]).read()
+        stderr_log = open(command_obj.log_files["stderr"]).read()
+        command = " ".join(command_obj.command)
+        error_message = "Error executing: '%s':\n" % command
+        if stdout_log.strip():
+            error_message += "\nStdout:\n%s\n" % stdout_log
+        if stderr_log.strip():
+            error_message += "\nStderr:\n%s\n" % stderr_log
+        raise RuntimeError(error_message) from e
+
+
+def get_lower_level_group(
+    api, top_level_kwargs: Dict, _type: str, deployer_kwargs: Dict
+):
+    if _type is None:
+        raise ValueError(
+            "ConcreteDeployer doesn't have a 'TYPE' to target. Please use a sub-class of ConcreteDeployer."
+        )
+    return getattr(api(**top_level_kwargs), _type)(**deployer_kwargs)
+
+
+class Deployer(object):
+    methods = {}
+
+    def __init__(
+        self,
+        flow_file: str,
+        show_output: bool = False,
+        profile: Optional[str] = None,
+        env: Optional[Dict] = None,
+        cwd: Optional[str] = None,
+        **kwargs
+    ):
+        self.flow_file = flow_file
+        self.show_output = show_output
+        self.profile = profile
+        self.env = env
+        self.cwd = cwd
+        self.top_level_kwargs = kwargs
+
+        from metaflow.plugins import CONCRETE_DEPLOYER_PROVIDERS
+
+        for provider_class in CONCRETE_DEPLOYER_PROVIDERS:
+            method_name = provider_class.TYPE.replace("-", "_")
+            Deployer.methods[method_name] = provider_class
+            setattr(Deployer, method_name, self.make_function(provider_class))
+
+    def make_function(self, deployer_class):
+        def f(self, **deployer_kwargs):
+            return deployer_class(
+                deployer_kwargs=deployer_kwargs,
+                flow_file=self.flow_file,
+                show_output=self.show_output,
+                profile=self.profile,
+                env=self.env,
+                cwd=self.cwd,
+                **self.top_level_kwargs
+            )
+
+        return f
+
+
+class TriggeredRun(object):
+    def __init__(
+        self,
+        deployer: "ConcreteDeployer",
+        content: str,
+    ):
+        self.deployer = deployer
+        content_json = json.loads(content)
+        self.metadata_for_flow = content_json.get("metadata")
+        self.pathspec = content_json.get("pathspec")
+        self.name = content_json.get("name")
+
+    def __contains__(self, key: str):
+        return getattr(self, key, None) is not None
+
+    def get(self, key: str, default=None) -> Optional[Any]:
+        return getattr(self, key, default)
+
+    def _update_env(self, env):
```

Comment:
nit: rename to something like `_enrich_object` (I liked that enrich term). _update_env can be a bit confusing here -- we are actually updating the object itself.